### PR TITLE
refactor secondary url caching

### PR DESCRIPTION
**Pull request by murali-shris**
**- What I did**
- Refactored secondary url cache

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by murali-shris** [comment]
TODO:
* exception handling when entry in cache expires and lookup fails - to be taken up later
* defaultCacheDuration and refreshCacheNow - temporarily commented.have to include in refactored code - to be taken up later
* throw exceptions from _findSecondary  - DONE

File: at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        // cache has not yet expired
+        return false;
+      }
+    } else {
+      // cache is empty
+      return true;
+    }
+  }
+
+  static String stripAtSignFromAtSign(String atSign) {
```
Comment (VJag) [inline at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart:72]: I thought we had a utility method already somewhere to do this. I believe this method can be moved to a reusable place.

File: at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    if (atSign.startsWith('@')) {
+      atSign = atSign.replaceFirst('@', '');
+    }
+    return atSign;
+  }
+
+  static String getNotFoundExceptionMessage(String atSign) {
+    return 'Unable to find secondary address for atSign:$atSign';
```
Comment (VJag) [inline at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart:80]: @murali-shris @kalluriramkumar We have to make the exception messages easily discoverable across the SDK. We will revisit this as part of the exception analysis task.

File: at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+          socket.destroy();
+          AtSignLogger('AtLookup').finer(
+              'AtLookup.findSecondary got answer: $secondary and closing connection');
+          return response;
+        }
+      }
+      // .. and close the socket
+      await socket.flush();
+      socket.destroy();
+      throw Exception('AtLookup.findSecondary timed out');
```
Comment (VJag) [inline at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart]: According to our exception hierarchy, I recon this corresponds to TimeoutException. We will revisit this in a different PR.

File: at_lookup/test/secondary_address_cache_test.dart [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
 import 'package:mocktail/mocktail.dart';
 
-class MockSecondaryFinder extends Mock implements SecondaryAddressFinder {}
+class MockSecondaryFinder extends Mock implements SecondaryUrlFinder {}
 
 void main() async {
   String rootDomain = 'root.atsign.unit.tests';
   int rootPort = 64;
-  SecondaryAddressFinder mockSecondaryFinder = MockSecondaryFinder();
+  SecondaryUrlFinder mockSecondaryFinder = MockSecondaryFinder();
```
Comment (VJag) [inline at_lookup/test/secondary_address_cache_test.dart:12]: Very very nice to see mock in action. Abstracting out external dependency into SecondaryUrlFinder makes this possible. Good work Murali. Yayyyy.

**Review by VJag — APPROVED** [review]
Minor comments. Good job.

File: at_lookup/test/secondary_address_cache_test.dart (hunk 4 above)
Comment (murali-shris) [reply at_lookup/test/secondary_address_cache_test.dart:12]: the tests were written  by Gary as a part of initial checkin

File: at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart (hunk 3 above)
Comment (murali-shris) [reply at_lookup/lib/src/cache/cacheable_secondary_address_finder.dart]: made the change to throw exact exceptions

**Pull request merged**